RecoLocalTracker/SiPhase2VectorHitBuilder/src/VectorHitBuilderAlgorithmBase.cc

```diff
@@ -0,0 +1,132 @@
+#include "Geometry/CommonDetUnit/interface/GeomDet.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
+
+VectorHitBuilderAlgorithmBase::VectorHitBuilderAlgorithmBase(const edm::ParameterSet& conf)
+    : nMaxVHforeachStack(conf.getParameter<int>("maxVectorHitsInAStack")),
+      barrelCut(conf.getParameter<std::vector<double> >("BarrelCut")),
+      endcapCut(conf.getParameter<std::vector<double> >("EndcapCut")),
+      cpeTag_(conf.getParameter<edm::ESInputTag>("CPE")) {}
+
+void VectorHitBuilderAlgorithmBase::initialize(const edm::EventSetup& es) {
+  // get the geometry and topology
+  edm::ESHandle<TrackerGeometry> geomHandle;
+  es.get<TrackerDigiGeometryRecord>().get(geomHandle);
+  initTkGeom(geomHandle);
+
+  edm::ESHandle<TrackerTopology> tTopoHandle;
+  es.get<TrackerTopologyRcd>().get(tTopoHandle);
+  initTkTopo(tTopoHandle);
+
+  // load the cpe via the eventsetup
+  edm::ESHandle<ClusterParameterEstimator<Phase2TrackerCluster1D> > cpeHandle;
+  es.get<TkPhase2OTCPERecord>().get(cpeTag_, cpeHandle);
+  initCpe(cpeHandle.product());
+}
+
+void VectorHitBuilderAlgorithmBase::initTkGeom(edm::ESHandle<TrackerGeometry> tkGeomHandle) {
+  theTkGeom = tkGeomHandle.product();
+}
+void VectorHitBuilderAlgorithmBase::initTkTopo(edm::ESHandle<TrackerTopology> tkTopoHandle) {
+  theTkTopo = tkTopoHandle.product();
+}
+void VectorHitBuilderAlgorithmBase::initCpe(const ClusterParameterEstimator<Phase2TrackerCluster1D>* cpeProd) {
+  cpe = cpeProd;
+}
+
+double VectorHitBuilderAlgorithmBase::computeParallaxCorrection(const PixelGeomDetUnit*& geomDetUnit_low,
+                                                                const Point3DBase<float, LocalTag>& lPosClu_low,
+                                                                const PixelGeomDetUnit*& geomDetUnit_upp,
+                                                                const Point3DBase<float, LocalTag>& lPosClu_upp) {
+  double parallCorr = 0.0;
+  Global3DPoint origin(0, 0, 0);
+  Global3DPoint gPosClu_low = geomDetUnit_low->surface().toGlobal(lPosClu_low);
+  GlobalVector gV = gPosClu_low - origin;
+  LogTrace("VectorHitsBuilderValidation") << " global vector passing to the origin:" << gV;
+
+  LocalVector lV = geomDetUnit_low->surface().toLocal(gV);
+  LogTrace("VectorHitsBuilderValidation") << " local vector passing to the origin (in low sor):" << lV;
+  LocalVector lV_norm = lV / lV.z();
+  LogTrace("VectorHitsBuilderValidation") << " normalized local vector passing to the origin (in low sor):" << lV_norm;
+
+  Global3DPoint gPosClu_upp = geomDetUnit_upp->surface().toGlobal(lPosClu_upp);
+  Local3DPoint lPosClu_uppInLow = geomDetUnit_low->surface().toLocal(gPosClu_upp);
+  parallCorr = lV_norm.x() * lPosClu_uppInLow.z();
+
+  return parallCorr;
+}
+
+void VectorHitBuilderAlgorithmBase::printClusters(const edmNew::DetSetVector<Phase2TrackerCluster1D>& clusters) {
+  int nCluster = 0;
+  int numberOfDSV = 0;
```

Comment:
since only LogTrace is used, better wrap the implementation in ```#ifdef EDM_ML_DEBUG```

Reply:
Done